src/coreclr/jit/morph.cpp

```diff
@@ -12443,6 +12443,34 @@ void Compiler::fgAssertionGen(GenTree* tree)
 #endif
     };
 
+    // If this tree creates an assignment of 0 or 1 to an int local, also create a [0..1] subrange
+    // assertion for that local, in case this local is used as a bool.
+    //
+    auto addImpliedBoolSubrangeAssertion = [=](AssertionIndex index, ASSERT_TP assertions) {
+        AssertionDsc* const assertion = optGetAssertion(index);
+        if ((assertion->assertionKind == OAK_EQUAL) && (assertion->op1.kind == O1K_LCLVAR) &&
+            (assertion->op2.kind == O2K_CONST_INT))
+        {
+            ssize_t iconVal = assertion->op2.u1.iconVal;
+            if ((iconVal == 0) || (iconVal == 1))
+            {
+                AssertionDsc extraAssertion   = {OAK_SUBRANGE};
+                extraAssertion.op1.kind       = O1K_LCLVAR;
+                extraAssertion.op1.lcl.lclNum = assertion->op1.lcl.lclNum;
+                extraAssertion.op2.kind       = O2K_SUBRANGE;
+                extraAssertion.op2.u2         = IntegralRange(SymbolicIntegerValue::Zero, SymbolicIntegerValue::One);
+
+                AssertionIndex extraIndex = optFinalizeCreatingAssertion(&extraAssertion);
+                if (extraIndex != NO_ASSERTION_INDEX)
+                {
+                    unsigned const bvIndex = extraIndex - 1;
+                    BitVecOps::AddElemD(apTraits, assertions, bvIndex);
+                    announce(extraIndex, "[bool range] ");
+                }
+            }
+        }
+    };
```

Comment:
@AndyAyersMS I am a bit confused - if a local already have `lcl = 1` (O1K_LCLVAR OAK_EQUAL O2K_CONST_INT) assertion, why does it need a separate `OAK_SUBRANGE` one?

Reply:
AP merge strategy is very simple, it just intersects bit vectors. So the merge of `X=1` from one path and `X=[0,1]` from another is the empty set, since those are different assertions, instead of the more reasonable `X=[0,1]`. 

A tricker case might be `X=0` on one path and `X=1` on another.

Adding merge smarts to AP likely would slow things down quite a bit (though we could try it as an alternative). But by adding a partially redundant `X=[0,1]` whenever we have `X=0` or `X=1` things just work out naturally, and the number of extra assertions isn't too bad.

Reply:
Ah, I see, makes sense then